### DISCUSSION TITLE
Is there some way to add data to composer.lock?

### DISCUSSION THREAD

**Discussion by cweagans** [composer/composer#11295] [Q&A / Support] [answered]
I'm working on the 2.0.0 release of https://github.com/cweagans/composer-patches in https://github.com/cweagans/composer-patches/pull/447.

I'd like to add a list of applied patches to each package in the lock file, and then when installing again in the future, only use the patches identified by the lock file.

Relevant lines are:

* https://github.com/cweagans/composer-patches/blob/9c1ea57a22fc17e8b0a9acd8bde7271ced9aa869/src/Plugin/Patches.php#L161-L184
* https://github.com/cweagans/composer-patches/blob/9c1ea57a22fc17e8b0a9acd8bde7271ced9aa869/src/Plugin/Patches.php#L284

The current version of the plugin writes the list of installed patches into installed.json, but I want version 2 to be a bit more rigorous about reproducible project state. Is there some way to add my own information to composer.lock or is this not a good idea?

**Comment by cweagans** [comment]
(cc @danepowell - maybe you know from your work on other composer plugins?)

**Comment by Seldaek** [comment, marked as answer]
You could technically patch the lock file but it is possibly you'd get clashes with other things and get your data overwritten. I'm not sure.

IMO it'd be better to take some approach similar to symfony/flex which adds a symfony.lock to keep track of stuff. If it provides value I'd expect people to be happy dealing with a patches.lock. 

The other option would be to dispatch an event in Composer when the lock file is dumped.

But ehh now that I think of it (sorry just keeping the stream of consciousness dump above in case this one doesn't pan out), maybe you could do resolution at some point and store this on cweagans/composer-patches's package instance inside extra data. So let's give a more concrete example:

- On initial install, the plugin isn't yet available/loaded when lock file is written, so at that point when the plugin gets installed you could collect patches and whatever data you need to store you would read the lock file, update your package's extra key and write again. Probably I'd also write the content-hash from the top level into extra as well so you can detect if something changed
- On a further install on InstallerEvents::PRE_OPERATIONS_EXEC you could read the lock file to see if the list of locked patches is there, if yes you can reuse it (unless the content-hash isn't matching anymore then I am not sure what you want to do but probably you'd need to recompute?)

One complication to keep in mind is if there is no vendor dir and someone runs a partial update, some of the lock file will be updated but your extra data will persist, that's why IMO you need the content-hash there and if it changed you need to consider the whole thing invalid otherwise you'll end up with unexpected behaviors.

**Reply by cweagans** [reply]
Oh, interesting. I hadn't thought about having a separate lock file. That might be a more straightforward path.

Thank you for the information!!
